lazyflow/operators/ioOperators/opExportMultipageTiff.py

```diff
@@ -149,24 +130,43 @@ def _batch_size(self) -> int:
         return min(batch_size, npages)
 
 
-class _NdBuf:
-    """Store ND-indexed items and give them back in the strict ND-ascending order."""
+class _ReqSyncBuf:
+    """Interate over items in ND-ascending order, that are set in arbitrary order
+
 
-    def __init__(self, shape: Iterable[int]):
+    TODO: add tests
+    TODO: include _page_rois, as this class relies on using the same logic for
+          keys
+    """
+
+    def __init__(self, shape, n_rois):
+        self.sig = threading.Event()
```

Comment:
I don't like this solution. It should be possible to redesign this using a blocking queue.

Reply:
I have to admit that I felt a bit weary when writing this concept is the main reason I wanted a review at this point. I have canvassed the stdlib before going here. So more concrete suggestions are highly welcome.

Reply:
Also this will most definitely block if there is an uncaught exception in some worker thread. :/

Reply:
Maybe something like this:

```python
import queue
from typing import Any, Iterable, MutableMapping

import numpy


class _NdBuf:
    """Store ND-indexed items and give them back in the strict ND-ascending order."""

    def __init__(self, shape: Iterable[int]):
        self._items = {}
        self._shape = tuple(shape)

    def __setitem__(self, key: Iterable[int], value: Any):
        i = numpy.ravel_multi_index(key, self._shape)
        self._items[i] = value

    def __iter__(self):
        return _NdBufIter(self._items)


class _NdBufIter:
    def __init__(self, items: MutableMapping[int, Any]):
        self._items = items
        self._i = 0

    def __iter__(self):
        return self

    def __next__(self):
        try:
            item = self._items.pop(self._i)
        except KeyError:
            raise StopIteration from None
        self._i += 1
        return item


def as_completed(buf: _NdBuf, q: queue.Queue):
    it = iter(buf)

    while True:
        kv = q.get()
        if kv is None:
            q.task_done()
            return

        buf[kv[0]] = kv[1]
        for item in it:
            q.task_done()
            yield item


def example():
    import threading
    import time
    
    buf = _NdBuf((2, 3))
    q = queue.Queue()
    
    def worker():
        # This is some order in which our blocks will be arriving.
        # Node that the first 3 items are out of order.
        for k in (0, 1), (0, 2), (0, 0), (1, 0), (1, 1), (1, 2):
            # Just some int value to send; we will be using an array here.
            v = 42 + numpy.ravel_multi_index(k, (2, 3))
            # Simulate work.
            time.sleep(0.5)
            print(f"processing {k} -> {v}")
            time.sleep(0.5)
            q.put((k, v))
        # Tell the other end to stop.
        q.put(None)
    
    threading.Thread(target=worker).start()
    
    for item in as_completed(buf, q):
        print(f"received {item}")


example()
```

Output:

```
processing (0, 1) -> 43
processing (0, 2) -> 44
processing (0, 0) -> 42
received 42
received 43
received 44
processing (1, 0) -> 45
received 45
processing (1, 1) -> 46
received 46
processing (1, 2) -> 47
received 47
```

Reply:
interesting. I think I'll explore a bit modifying `RoiRequestBatch`. Using it right now in this ordered manner seems like a workaround.

Reply:
Maybe it makes sense to add `join` method to `RoiRequestsBatch` to await the completion of all pending requests.

Reply:
while I can see that a `join` method can be useful in some cases, I don't think this is want we want to use here - data can be potentially be bigger than ram, so we'd want to write/discard data as quickly as possible.